.github/workflows/CICD.yml

```diff
@@ -1,18 +1,124 @@
 name: CICD
-
-on: [push, pull_request]
-
-env:
-  CARGO_TERM_COLOR: always
-
+on: [pull_request, push]
 jobs:
-  build:
-
+  test-lib:
+    name: Test library
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-latest, windows-latest]
+        toolchain: [stable, nightly]
+    continue-on-error: true
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Rust
+        run: |
+          rustup toolchain add ${{ matrix.toolchain }} --component llvm-tools-preview
+          rustup override set ${{ matrix.toolchain }}
+      - name: Configure cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            target/
+          key: test-${{ runner.os }}-cargo-${{ matrix.toolchain }}-${{ hashFiles('**/Cargo.lock') }}
+      - name: Test
+        run: cargo test --lib
+  test-coverage:
+    name: Test coverage (gcc-${{ matrix.gcc }}, clang-${{ matrix.clang }})
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        gcc: [10, 11]
+        clang: [11, 12, 13]
```

Comment:
There's no need to test all combinations, we just need to make sure there is at least one task for each gcc version and each clang version.

E.g.:
- gcc 10, clang 11
- gcc 11, clang 12
- gcc 10, clang 13

Instead of:
- gcc 10, clang 11
- gcc 11, clang 11
- gcc 10, clang 12
- gcc 11, clang 12
- gcc 10, clang 13
- gcc 11, clang 13